### PR TITLE
As a Superadmin, I want to filter the Volunteers index table

### PR DESCRIPTION
**Pull request by casaper**
[Link to story in trello](https://trello.com/c/O8YvU3Jj/16-as-a-superadmin-i-want-to-filter-the-volunteers-index-table-sp-3)

- [x] Story under test
- [x] Mobile works
- [x] Seeds created
- [x] Translated

Review by Team
- [ ] Reviewed by tnt
- [x] Reviewed by cla
- [ ] Reviewed by vok

Review by Coach
- [x] Reviewed by sta

### CODE REVIEW

File: app/models/concerns/gender_collection.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+module GenderCollection
```
Comment (chrysalag) [inline app/models/concerns/gender_collection.rb]: maybe all collection could be in one concern?

File: app/views/volunteers/show.html.slim [hunk 2]
```diff
@@ -7,83 +7,103 @@
...
-          tr
-            td= t_attr(:state)
-            td= @volunteer.state
-          tr
-            td= t_attr(:duration)
-            td= t("duration.#{@volunteer.duration}") unless @volunteer.duration.blank?
-          tr
-            td= t_attr(:region)
-            td= t("region.#{@volunteer.region}") unless @volunteer.region.blank?
+          - if @volunteer.date_of_birth.present?
```
Comment (chrysalag) [inline app/views/volunteers/show.html.slim]: What is with having in the view also the unfilled fields? Also, it is likely that all these fields will become required. At least, this what I understood from @kitsane .

File: app/models/volunteer.rb [hunk 3]
```diff
@@ -31,4 +35,8 @@ def self.group_accompaniment
   def self.human_boolean(boolean)
     boolean ? I18n.t('simple_form.yes') : I18n.t('simple_form.no')
   end
+
+  def to_s
```
Comment (chrysalag) [inline app/models/volunteer.rb]: why is this needed?

File: db/schema.rb [hunk 4]
```diff
@@ -192,7 +192,7 @@
     t.text "strengths"
     t.text "skills"
     t.text "interests"
-    t.string "state", default: "interested/registered"
+    t.string "state", default: "registered"
```
Comment (chrysalag) [inline db/schema.rb]: It's a minor thing in this case, but we've said we don't edit old migrations.

File: app/views/volunteers/show.html.slim (hunk 2 above)
Comment (casaper) [inline app/views/volunteers/show.html.slim]: For any that aren't required it should be like this, and if they're required they'll always be there. So no difference...

File: app/models/volunteer.rb (hunk 3 above)
Comment (casaper) [inline app/models/volunteer.rb]: needed for confirmation message and many other things. this is standard practice...

File: app/models/concerns/gender_collection.rb (hunk 1 above)
Comment (casaper) [inline app/models/concerns/gender_collection.rb]: Keep things seperated, especially as one or the other could be shifted into the contact model later on.

File: db/schema.rb (hunk 4 above)
Comment (casaper) [inline db/schema.rb]: this isn't a migration

Comment (chrysalag) [inline db/schema.rb]: Yes, but it is generated through the migrations. Travis currently just loads the schema, but when someone runs db:migrate, this line of the schema will change again.

File: db/migrate/20170519173558_change_volunteer_state_and_email.rb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+class ChangeVolunteerStateAndEmail < ActiveRecord::Migration[5.1]
+  def change
+    change_column_default :volunteers, :state, 'registered'
+    add_column :volunteers, :active, :boolean
+    add_index :volunteers, :email, unique: true, where: 'active'
```
Comment (kitsane) [inline db/migrate/20170519173558_change_volunteer_state_and_email.rb]: This should be in a seperate pull request

File: app/views/volunteers/show.html.slim (hunk 2 above)
Comment (kitsane) [inline app/views/volunteers/show.html.slim]: This is part of a UX question, i prefer to see the empty fields, because when i know how the view looks and then don't see the keys, i get the feeling that something is wrong.

Second is to check every field with `.present?` is a lot of repetition, this could be done with a helper method

File: Gemfile [hunk 6]
```diff
@@ -26,6 +27,8 @@ gem 'paranoia'
 gem 'pg'
 gem 'puma'
 gem 'pundit'
+gem 'rails-i18n', '~> 5.0.0'
```
Comment (kitsane) [inline Gemfile]: Only specify the version when it is really needed

File: Gemfile [hunk 7]
```diff
@@ -26,6 +27,8 @@ gem 'paranoia'
...
+gem 'ransack', github: 'activerecord-hackery/ransack'
```
Comment (kitsane) [inline Gemfile]: Do we need the github version of the gem? I whould prefer the rubygems version for stability reasons

File: app/views/volunteers/index.html.slim [hunk 8]
```diff
@@ -1,25 +1,24 @@
...
-        th= t_attr(:email)
-        th
-        th
-        th
+        th= sort_link @q, :first_name, t_attr(:full_name)
+        th= sort_link @q, :date_of_birth
+        th= sort_link @q, :nationality
+        th= sort_link @q, :created_at
+        th= sort_link @q, :gender
+        th= render 'state_select'
```
Comment (kitsane) [inline app/views/volunteers/index.html.slim]: I whould prefer, the table headers to be only used for sorting, same like the other ths

File: db/schema.rb (hunk 4 above)
Comment (casaper) [inline db/schema.rb]: nope. Just look at my migration.

File: Gemfile (hunk 7 above)
Comment (casaper) [inline Gemfile]: In your Gemfile, for the last officially released gem:
```ruby
gem 'ransack'
```

If you would like to use the latest updates (recommended), use the master branch:
```ruby
gem 'ransack', github: 'activerecord-hackery/ransack'
```

Comment (casaper) [inline Gemfile]: they recommend it, thats why I did it. But I havn't tryed without tho

File: app/views/volunteers/index.html.slim (hunk 8 above)
Comment (casaper) [inline app/views/volunteers/index.html.slim]: ok.

File: Gemfile (hunk 6 above)
Comment (casaper) [inline Gemfile]: see https://github.com/svenfuchs/rails-i18n#gem-installation

File: app/policies/volunteer_policy.rb [hunk 9]
```diff
@@ -0,0 +1,12 @@
+class VolunteerPolicy < ApplicationPolicy
+  attr_reader :user, :volunteer
+
+  def initialize(user, volunteer)
+    @user = user
+    @volunteer = volunteer
+  end
+
+  def access_state?
```
Comment (kitsane) [inline app/policies/volunteer_policy.rb]: This method can be dropped, as the same is defined in the ApplicationPolicy

File: app/controllers/volunteers_controller.rb [hunk 10]
```diff
@@ -1,9 +1,11 @@
 class VolunteersController < ApplicationController
   include NestedAttributes
+  include VolunteerAttributes
   before_action :set_volunteer, only: [:show, :edit, :update, :destroy]
 
   def index
-    @volunteers = Volunteer.all
+    @q = Volunteer.ransack(params[:q] || { state_cont: 'registered' })
```
Comment (kitsane) [inline app/controllers/volunteers_controller.rb]: It should still be possible to access the index without params:
```ruby
if params[:q].present?
  @q = Volunteer.ransack(params[:q])
  @volunteers = @q.result
else
  @volunteers = Volunteer.all
end
```

File: app/views/volunteers/_state_select.slim [hunk 11]
```diff
@@ -0,0 +1,16 @@
+ruby:
```
Comment (kitsane) [inline app/views/volunteers/_state_select.slim]: Please no inline `ruby` in views

File: app/views/volunteers/_state_select.slim [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+  scope = {scope: [:simple_form, :options, :volunteer, :state]}
```
Comment (kitsane) [inline app/views/volunteers/_state_select.slim]: Is this still needed? There is no form here

File: app/views/volunteers/_state_select.slim [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+  size ||= 'xs'
+  descriptive ||= false
```
Comment (kitsane) [inline app/views/volunteers/_state_select.slim]: Drop the `size` and `descriptive` options

File: app/views/volunteers/_state_select.slim [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+  state_filter = @q.base.conditions.first.values.first.value if @q.base.conditions.first
+  label = "#{t_attr(:state)} "
+  label += ": #{t(state_filter, scope)} " if descriptive && state_filter.present?
```
Comment (kitsane) [inline app/views/volunteers/_state_select.slim]: Just make the label to `states` for now

File: app/views/volunteers/_state_select.slim [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+.btn-group
+  button.btn.btn-default.dropdown-toggle type='button' data-toggle='dropdown' aria-label="#{t_attr(:state)}" aria-haspopup='true' aria-expanded='false' class="btn-#{size}"
+    = label
+    span.caret aria-hidden='true'
+  ul.dropdown-menu
+    li
+      = link_to t('all'), "#{volunteers_path}?q[state_cont]="
```
Comment (kitsane) [inline app/views/volunteers/_state_select.slim]: With the changes in the controller use:
```ruby
ul.dropdown-menu
  li
    = link_to t('all'), volunteers_path
  - Volunteer.state_collection.each do |state|
    li= link_to t(".states.#{state.to_s}"), volunteers_path(q: { state_cont: state })"
```

File: app/views/volunteers/index.html.slim [hunk 16]
```diff
@@ -1,23 +1,24 @@
 .row
   .col-xs-12
     h1= t('volunteers')
-
+.row
+  .col-xs-12= render 'state_select', size: 'md', descriptive: true
```
Comment (kitsane) [inline app/views/volunteers/index.html.slim]: remove `size` and `descriptive`

File: config/application.rb [hunk 17]
```diff
@@ -1,5 +1,6 @@
 require_relative 'boot'
-
+require File.expand_path('../boot', __FILE__)
+ENV['RANSACK_FORM_BUILDER'] = '::SimpleForm::FormBuilder'
```
Comment (kitsane) [inline config/application.rb]: Is this still needed?

File: config/locales/de.yml [hunk 18]
```diff
@@ -448,11 +441,13 @@ de:
   users: Benutzer
   voluntary_platform: Freiwilligen Plattform
   volunteer_application: Freiwillige/n Bewerbung
+  volunteer_application: Freiwillige/n Bewerbung
```
Comment (nuritnt) [inline config/locales/de.yml]: duplicate?

File: config/locales/de.yml [hunk 19]
```diff
@@ -448,11 +441,13 @@ de:
...
   volunteer_created: Freiwillige wurde erfolgreich erstellt.
-  volunteer_destroyed: 'Freiwillige wurde erfolgreich gelöscht.'
+  volunteer_destroyed: Freiwillige wurde erfolgreich gelöscht.
   volunteer_updated: Freiwillige wurde erfolgreich aktualisiert.
   volunteers: Freiwillige
   wanna_apply: Möchten Sie sich als Freiwillige/n bewerben?
+  wanna_apply: Möchten Sie sich als Freiwillige/n bewerben?
```
Comment (nuritnt) [inline config/locales/de.yml]: duplicate?

File: config/locales/en.yml [hunk 20]
```diff
@@ -50,6 +41,8 @@ en:
       volunteer_application: Volunteer Application
   add_language: Add language
   add_relative: Add family member
+  all: all
+  application_submitted: 'Your application with the email %{email} has been successfully sent.'
```
Comment (nuritnt) [inline config/locales/en.yml]: duplicate?

File: config/locales/en.yml [hunk 21]
```diff
@@ -295,11 +298,13 @@ en:
   users: Users
   voluntary_platform: Voluntaries Platform
   volunteer_application: Volunteer Application
-  volunteer_created: 'Volunteer was successfully created.'
-  volunteer_destroyed: 'Volunteer was successfully deleted.'
-  volunteer_updated: 'Volunteer was successfully updated.'
+  volunteer_application: Volunteer Application
```
Comment (nuritnt) [inline config/locales/en.yml]: duplicate?

File: config/locales/en.yml [hunk 22]
```diff
@@ -295,11 +298,13 @@ en:
...
+  volunteer_created: Volunteer was successfully created.
+  volunteer_destroyed: Volunteer was successfully deleted.
+  volunteer_updated: Volunteer was successfully updated.
   volunteers: Volunteers
   wanna_apply: Do you want to apply as a volunteer?
+  wanna_apply: Do you want to apply as a volunteer?
```
Comment (nuritnt) [inline config/locales/en.yml]: ddddddduplicate?

File: db/schema.rb (hunk 4 above)
Comment (casaper) [inline db/schema.rb]: I've just seen what you mean, because you've also set a different default `interested` in your branch. But I adopted the same state collection as used in the clients model as well, and there it is registered, not interested.

File: Gemfile (hunk 7 above)
Comment (casaper) [inline Gemfile]: But how come they recommend using theyre master version?

File: app/controllers/volunteers_controller.rb (hunk 10 above)
Comment (casaper) [inline app/controllers/volunteers_controller.rb]: the requirement is that the default filter is like this. And it is still possible to see all. If I do what you suggest, the acceptance criteria is not going to be fullfilled.

File: Gemfile (hunk 7 above)
Comment (kitsane) [inline Gemfile]: Maybe it is one of the gems that is stable on master.

File: app/controllers/volunteers_controller.rb (hunk 10 above)
Comment (kitsane) [inline app/controllers/volunteers_controller.rb]: Ok, please drop the acceptance criteria

File: Gemfile (hunk 6 above)
Comment (kitsane) [inline Gemfile]: The gemspec file will ensure, that bundler takes the right version: https://github.com/svenfuchs/rails-i18n/blob/master/rails-i18n.gemspec#L22

File: app/policies/volunteer_policy.rb (hunk 9 above)
Comment (casaper) [inline app/policies/volunteer_policy.rb]: Where? I don't find anything there that does the same?

File: app/controllers/volunteers_controller.rb (hunk 10 above)
Comment (casaper) [inline app/controllers/volunteers_controller.rb]: Well, fine. No problem. But I save this as a credit for bailing the process next time I like to ;-)

File: config/application.rb (hunk 17 above)
Comment (casaper) [inline config/application.rb]: not sure. I was following the Ransack manual. But maybe it is not necessary.

File: app/policies/volunteer_policy.rb (hunk 9 above)
Comment (kitsane) [inline app/policies/volunteer_policy.rb]: I looked in the wrong policy
